controller/array_action/array_mediator_svc.py

```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
             logger.debug("can not found host by using initiators: {0} ".format(initiators))
             raise controller_errors.HostNotFoundError(initiators)
 
+    def _get_detailed_hosts_list(self):
+        logger.debug("Getting detailed hosts list on array {0}".format(self.endpoint))
+        hosts_list = self.client.svcinfo.lshost()
+        if not hosts_list:
+            return []
+
+        # for each host get its detailed info from array by sending batch of commands
+        detailed_hosts_list_cmd = self._get_detailed_hosts_list_cmd(hosts_list)
+        logger.debug("Sending getting detailed hosts list commands batch")
+        detailed_hosts_list_output, detailed_hosts_list_errors = self.client.send_raw_command(detailed_hosts_list_cmd)
+        detailed_hosts_list_as_string = bytes_to_string(detailed_hosts_list_output)
+        detailed_hosts_list_error_msg = self._get_formatted_hosts_list_error_msg(detailed_hosts_list_errors)
+        if not detailed_hosts_list_errors:
+            logger.error("Errors returned from getting detailed hosts list: {0}".format(detailed_hosts_list_error_msg))
+
+        logger.debug("Reading detailed hosts list commands batch response")
+        hosts_reader = SVCListResultsReader(detailed_hosts_list_as_string)
+        res = []
+        for host_details in hosts_reader:
+            host_id = host_details.get(HOST_ID_PARAM)
+            host_name = host_details.get(HOST_NAME_PARAM)
+            iscsi_names = host_details.get_as_list(HOST_ISCSI_NAMES_PARAM)
+            wwns = host_details.get_as_list(HOST_WWPNS_PARAM)
+            host = Host(host_id, host_name, iscsi_names, wwns)
+            res.append(host)
+        return res
+
+    def _get_detailed_hosts_list_cmd(self, host_list):
+        writer = StringIO()
+        for host in host_list:
+            host_id = host.get(HOST_ID_PARAM)
+            writer.write(LIST_HOSTS_CMD)
```

Comment:
I think something like
```
LIST_HOSTS_CMD_FORMAT = 'lshost {HOST_ID};'
writer.write(LIST_HOSTS_CMD_FORMAT.format(HOST_ID=host_id)
```
would be more readable and with the same "big O notation" efficiency

Reply:
it is still linear to the amount of hosts (and I assume "thousands" is only a theoretical case).
so how do you know that 1 short string construction requires more instructions than 4 "append"s?
did you compare the performance between the two?

Reply:
It doubles the time. It is not theoretical. We should support 4000 hosts.. 
Format passes over all characters and then  write passes again.

Reply:
why double? it should also do 3 times less appends.
anyway, you can also try something like `map(writer.write, ('lshost ', host_id, ';'))`

Reply:
It first "formats" pass for each char in "lshost 1" and then "writes to writer" for each char.in "lshost 1"

Reply:
```
import timeit
from io import StringIO

def _first(host_list):
	writer = StringIO()
	for host_id in host_list:
		writer.write('lshost')
		writer.write(' ')
		writer.write(host_id)
		writer.write(';')
	return writer.getvalue()


def _second(host_list):
    writer = StringIO()
    for host_id in host_list:
        writer.write('lshost {HOST_ID};'.format(HOST_ID=host_id))
    return writer.getvalue()


four_k = map(str, range(4000))

t1 = timeit.Timer("_first(four_k)", "from __main__ import _first, four_k")
first_time = t1.timeit(1)
print("1st took %fs" % (first_time,))

t2 = timeit.Timer("_second(four_k)", "from __main__ import _second, four_k")
second_time = t2.timeit(1)
print("2nd took %fs" % (second_time,))
```
1st took 0.005819s
2nd took 0.000017s

Reply:
I'm not sure I follow the theory.
did you try to compare their time?

Reply:
if you want to save the time comparisons, this is the fastest and most pythonic approach:
`return 'lshost {}'.format(';lshost '.join(host_list))`

Reply:
I don't understand from your last response why the writer is faster - is this about the `map` overhead?

Reply:
How return 'lshost {}'.format(';lshost '.join(host_list))
do the job?

Reply:
I'll use: LIST_HOSTS_CMD_FORMAT = 'lshost {HOST_ID};'
writer.write(LIST_HOSTS_CMD_FORMAT.format(HOST_ID=host_id)